FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_keep_alive/DemoTasks/KeepAliveMQTTExample.c

```diff
@@ -0,0 +1,773 @@
+static void prvMQTTProcessIncomingPublish( MQTTPublishInfo_t * pxPublishInfo );
+
+/**
+ * @brief This callback is invoked through an auto-reload software timer.
+ *
+ * Its responsibility is to send a PINGREQ packet if a PINGRESP is not pending
+ * and no control packets have been sent after some given interval.
+ *
+ * @param pxTimer The auto-reload software timer for handling keep alive.
+ */
+static void prvKeepAliveTimerCallback( TimerHandle_t pxTimer );
+
+/**
+ * @brief The application callback function for getting the incoming publish
+ * and incoming acks reported from the MQTT library.
+ *
+ * @param pxMQTTContext MQTT context pointer.
+ * @param pxPacketInfo Packet Info pointer for the incoming packet.
+ * @param usPacketIdentifier Packet identifier of the incoming packet.
+ * @param pxPublishInfo Deserialized publish info pointer for the incoming
+ * packet.
+ */
+static void prvEventCallback( MQTTContext_t * pxMQTTContext,
+                              MQTTPacketInfo_t * pxPacketInfo,
+                              uint16_t usPacketIdentifier,
+                              MQTTPublishInfo_t * pxPublishInfo );
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Static buffer used to hold MQTT messages being sent and received.
+ */
+static uint8_t ucSharedBuffer[ mqttexampleSHARED_BUFFER_SIZE ];
+
+/**
+ * @brief Static buffer used to hold PINGREQ messages to be sent.
+ */
+static uint8_t ucPingReqBuffer[ MQTT_PACKET_PINGREQ_SIZE ];
+
+/**
+ * @brief Global entry time into the application to use as a reference timestamp
+ * in the #prvGetTimeMs function. #prvGetTimeMs will always return the difference
+ * between the current time and the global entry time. This will reduce the chances
+ * of overflow for the 32 bit unsigned integer used for holding the timestamp.
+ */
+static uint32_t ulGlobalEntryTimeMs;
+
+/**
+ * @brief Packet Identifier generated when Subscribe request was sent to the broker;
+ * it is used to match received Subscribe ACK to the transmitted ACK.
+ */
+static uint16_t usSubscribePacketIdentifier;
+
+/**
+ * @brief Packet Identifier generated when Unsubscribe request was sent to the broker;
+ * it is used to match received Unsubscribe response to the transmitted unsubscribe
+ * request.
+ */
+static uint16_t usUnsubscribePacketIdentifier;
+
+/**
+ * @brief Timer to handle the MQTT keep-alive mechanism.
+ */
+static TimerHandle_t xKeepAliveTimer;
+
+/**
+ * @brief Storage space for xKeepAliveTimer.
+ */
+static StaticTimer_t xKeepAliveTimerBuffer;
+
+/**
+ * @brief Set to true when PINGREQ is sent then false when PINGRESP is received.
+ */
+static volatile bool xWaitingForPingResp = false;
+
+/**
+ * @brief The last time when a PINGREQ was sent over the network.
+ */
+static uint32_t ulPingReqSendTimeMs;
+
+/**
+ * @brief Timeout for a pending PINGRESP from the MQTT broker.
+ */
+static uint32_t ulPingRespTimeoutMs = mqttexampleKEEP_ALIVE_TIMEOUT_SECONDS * _MILLISECONDS_PER_SECOND;

```

Comment:
This seems too long. If the server didn't receive the ping, then it will have already closed the connection halfway before this interval is over.

Reply:
Set it to be as long as the timer interval